File: src/lib/agent/filter-engine.ts
import { asDate, asNumber, matchColumn, suggestColumns } from "./semantics";
import type {
  DataRow,
  FilterCondition,
  FilterGroup,
  FilterResult,
  FilterTrace,
} from "./types";
import { isGroup } from "./types";

function str(v: unknown): string {
  return v == null ? "" : String(v);
}

function pair(value: unknown): [unknown, unknown] {
  if (Array.isArray(value) && value.length >= 2) return [value[0], value[1]];
  throw new Error("between / year_between require a two-element [min, max] value.");
}

function eqCell(cell: unknown, value: unknown): boolean {
  if (typeof cell === "number" && typeof value === "number") return cell === value;
  if (typeof cell === "boolean") return cell === Boolean(value);
  return str(cell).toLowerCase() === str(value).toLowerCase();
}

function cellIn(cell: unknown, values: unknown[]): boolean {
  return values.some((v) => eqCell(cell, v));
}

function numCell(cell: unknown): number | null {
  return asNumber((cell as never) ?? null);
}

function dateCell(cell: unknown): Date | null {
  return asDate((cell as never) ?? null);
}

export function resolveGroup(group: FilterGroup, columns: string[]): FilterGroup {
  return {
    logic: group.logic,
    conditions: group.conditions.map((item) => {
      if (isGroup(item)) return resolveGroup(item, columns);
      const real = matchColumn(item.column, columns);
      if (!real) {
        const suggestions = suggestColumns(item.column, columns);
        const hint = suggestions.length ? ` Did you mean: ${suggestions.join(", ")}?` : "";
        throw new Error(`Unknown column '${item.column}'. Available: ${columns.join(", ")}.${hint}`);
      }
      return { ...item, column: real };
    }),
  };
}

function estimateSelectivity(rows: DataRow[], item: FilterCondition | FilterGroup): number {
  if (isGroup(item)) {
    const scores = item.conditions.map((c) => estimateSelectivity(rows, c));
    if (!scores.length) return 1;
    if (item.logic === "and") {
      return Math.max(scores.reduce((p, s) => p * Math.max(s, 1e-6), 1), 0);
    }
    return Math.min(
      1,
      1 - scores.reduce((p, s) => p * Math.max(1 - s, 0), 1),
    );
  }
  const n = Math.max(rows.length, 1);
  const uniq = new Set(rows.map((r) => str(r[item.column]))).size || 1;
  switch (item.op) {
    case "eq":
      return Math.min(1, 1 / uniq);
    case "ne":
      return Math.min(1, Math.max(0, 1 - 1 / uniq));
    case "in":
      return Math.min(1, (Array.isArray(item.value) ? item.value.length : 1) / uniq);
    case "not_in":
      return Math.min(1, Math.max(0, 1 - (Array.isArray(item.value) ? item.value.length : 1) / uniq));
    case "between":
    case "year_between":
      return 0.25;
    case "year_eq":
      return 0.2;
    case "gt":
    case "gte":
    case "lt":
    case "lte":
    case "date_before":
    case "date_after":
      return 0.4;
    case "contains":
    case "starts_with":
    case "ends_with":
    case "regex":
      return 0.3;
    case "is_null":
      return rows.filter((r) => r[item.column] == null || r[item.column] === "").length / n;
    case "not_null":
      return rows.filter((r) => r[item.column] != null && r[item.column] !== "").length / n;
    default:
      return 0.5;
  }
}

function matchCondition(row: DataRow, cond: FilterCondition): boolean {
  const cell = row[cond.column];
  const { op, value } = cond;
  if (op === "is_null") return cell == null || cell === "";
  if (op === "not_null") return cell != null && cell !== "";

  if (op === "year_eq" || op === "year_between" || op === "date_before" || op === "date_after") {
    const d = dateCell(cell);
    if (!d) return false;
    if (op === "year_eq") {
      const year = Number(Array.isArray(value) ? value[0] : value);
      return d.getUTCFullYear() === year;
    }
    if (op === "year_between") {
      const [lo, hi] = pair(value);
      const y = d.getUTCFullYear();
      return y >= Number(lo) && y <= Number(hi);
    }
    const other = asDate((Array.isArray(value) ? value[0] : value) as never);
    if (!other) return false;
    if (op === "date_before") return d.getTime() < other.getTime();
    return d.getTime() > other.getTime();
  }

  if (op === "eq") return eqCell(cell, value);
  if (op === "ne") return !eqCell(cell, value);
  if (op === "in") return cellIn(cell, Array.isArray(value) ? value : [value]);
  if (op === "not_in") return !cellIn(cell, Array.isArray(value) ? value : [value]);
  if (op === "contains") return str(cell).toLowerCase().includes(str(value).toLowerCase());
  if (op === "not_contains") return !str(cell).toLowerCase().includes(str(value).toLowerCase());
  if (op === "starts_with") return str(cell).toLowerCase().startsWith(str(value).toLowerCase());
  if (op === "ends_with") return str(cell).toLowerCase().endsWith(str(value).toLowerCase());
  if (op === "regex") {
    try {
      return new RegExp(str(value), "i").test(str(cell));
    } catch {
      return false;
    }
  }
  if (op === "between") {
    const [lo, hi] = pair(value);
    const n = numCell(cell);
    if (n === null) {
      const d = dateCell(cell);
      const a = asDate(lo as never);
      const b = asDate(hi as never);
      if (!d || !a || !b) return false;
      return d.getTime() >= a.getTime() && d.getTime() <= b.getTime();
    }
    return n >= Number(lo) && n <= Number(hi);
  }
  const n = numCell(cell);
  if (n === null) return false;
  const rhs = Number(value);
  if (!Number.isFinite(rhs)) return false;
  if (op === "gt") return n > rhs;
  if (op === "gte") return n >= rhs;
  if (op === "lt") return n < rhs;
  if (op === "lte") return n <= rhs;
  return false;
}

function applyGroup(rows: DataRow[], group: FilterGroup, path: string): { rows: DataRow[]; trace: FilterTrace[] } {
  if (!group.conditions.length) return { rows, trace: [] };

  if (group.logic === "or") {
    const seen = new Set<DataRow>();
    const out: DataRow[] = [];
    const trace: FilterTrace[] = [];
    for (let i = 0; i < group.conditions.length; i += 1) {
      const item = group.conditions[i]!;
      const child = isGroup(item)
        ? applyGroup(rows, item, `${path}.or[${i}]`)
        : { rows: rows.filter((r) => matchCondition(r, item)), trace: [] as FilterTrace[] };
      let added = 0;
      for (const r of child.rows) {
        if (!seen.has(r)) {
          seen.add(r);
          out.push(r);
          added += 1;
        }
      }
      trace.push({
        path: `${path}.or[${i}]`,
        logic: "or",
        predicate: item,
        addedRows: added,
        child: child.trace,
      });
    }
    return { rows: out, trace };
  }

  const ordered = [...group.conditions].sort(
    (a, b) => estimateSelectivity(rows, a) - estimateSelectivity(rows, b),
  );
  let working = rows;
  const trace: FilterTrace[] = [];
  for (let i = 0; i < ordered.length; i += 1) {
    const item = ordered[i]!;
    const before = working.length;
    const child = isGroup(item)
      ? applyGroup(working, item, `${path}.and[${i}]`)
      : { rows: working.filter((r) => matchCondition(r, item)), trace: [] as FilterTrace[] };
    working = child.rows;
    const after = working.length;
    const pred = isGroup(item) ? item : item;
    trace.push({
      path: `${path}.and[${i}]`,
      logic: "and",
      order: i + 1,
      predicate: pred,
      estimatedSelectivity: estimateSelectivity(rows, item),
      rowsBefore: before,
      rowsAfter: after,
      dropped: before - after,
      emptiedResult: before > 0 && after === 0,
      note:
        before > 0 && after === 0
          ? "This predicate removed every remaining row. Combining it with AND hid the rest of the data. Consider relaxing it or switching same-column values to OR."
          : "Applied after more selective predicates so remaining rows stay visible.",
      child: child.trace,
    });
  }
  return { rows: working, trace };
}

function humanSummary(nIn: number, nOut: number, trace: FilterTrace[]): string {
  const pct = nIn ? ((100 * nOut) / nIn).toFixed(1) : "0.0";
  const bits = [`Matched ${nOut.toLocaleString()} of ${nIn.toLocaleString()} rows (${pct}%).`];
  if (nOut === 0) bits.push("The result is empty — see the filter trace for which predicate removed the last rows.");
  const walk = (nodes: FilterTrace[]) => {
    for (const node of nodes) {
      if (node.logic === "and" && (node.dropped ?? 0) > 0 && node.predicate && !isGroup(node.predicate)) {
        const p = node.predicate;
        bits.push(
          `AND step ${node.order}: ${p.column} ${p.op} ${JSON.stringify(p.value)} dropped ${node.dropped} rows (${node.rowsBefore} → ${node.rowsAfter}).`,
        );
      }
      if (node.child?.length) walk(node.child);
    }
  };
  walk(trace);
  return bits.join(" ");
}

export function applyFilter(rows: DataRow[], plan: FilterGroup): FilterResult {
  const columns = rows.length ? Object.keys(rows[0] ?? {}) : [];
  const resolved = resolveGroup(plan, columns);
  const { rows: matched, trace } = applyGroup(rows, resolved, "root");
  return {
    rows: matched,
    nInput: rows.length,
    nMatched: matched.length,
    nDropped: rows.length - matched.length,
    plan: resolved,
    trace,
    empty: matched.length === 0,
    summary: humanSummary(rows.length, matched.length, trace),
  };
}

export function describePlan(plan: FilterGroup): string {
  const parts = plan.conditions.map((c) => {
    if (isGroup(c)) return `(${describePlan(c)})`;
    const val = Array.isArray(c.value) ? c.value.join("–") : String(c.value ?? "");
    return `${c.column} ${c.op} ${val}`.trim();
  });
  return parts.join(` ${plan.logic.toUpperCase()} `);
}

File: excel_data_agent/core/filter_engine.py
"""Deterministic pandas filter engine.

The LLM never writes pandas code. It emits a FilterGroup JSON document; this
module resolves column names against the real header and applies operators.
AND groups are executed in selectivity order so the first predicate that
collapses the set is visible — that is the hierarchy of data visibility.
"""

from __future__ import annotations

from typing import Any

import pandas as pd

from excel_data_agent.core.column_semantics import coerce_value, match_column, suggest_columns
from excel_data_agent.core.hierarchy import estimate_selectivity, order_and_children
from excel_data_agent.core.types import FilterCondition, FilterGroup, json_safe, parse_filter_plan


def apply_plan(
    df: pd.DataFrame,
    plan: FilterGroup | dict[str, Any] | str,
) -> dict[str, Any]:
    group = parse_filter_plan(plan)
    columns = [str(c) for c in df.columns]
    resolved = _resolve_group(group, columns)
    mask, trace = _apply_group(df, resolved, path="root")
    filtered = df.loc[mask].copy()
    return {
        "status": "success",
        "n_input": int(len(df)),
        "n_matched": int(len(filtered)),
        "n_dropped": int(len(df) - len(filtered)),
        "plan": resolved.to_dict(),
        "trace": trace,
        "empty": bool(filtered.empty),
        "frame": filtered,
    }


def _resolve_group(group: FilterGroup, columns: list[str]) -> FilterGroup:
    resolved: list[FilterCondition | FilterGroup] = []
    for item in group.conditions:
        if isinstance(item, FilterGroup):
            resolved.append(_resolve_group(item, columns))
            continue
        real = match_column(item.column, columns)
        if real is None:
            suggestions = suggest_columns(item.column, columns)
            hint = f" Did you mean: {', '.join(suggestions)}?" if suggestions else ""
            raise ValueError(
                f"Unknown column '{item.column}'. Available columns: {', '.join(columns)}.{hint}"
            )
        resolved.append(FilterCondition(column=real, op=item.op, value=item.value))
    return FilterGroup(logic=group.logic, conditions=resolved)


def _apply_group(
    df: pd.DataFrame, group: FilterGroup, *, path: str
) -> tuple[pd.Series, list[dict[str, Any]]]:
    if not group.conditions:
        return pd.Series(True, index=df.index), []

    if group.logic == "or":
        mask = pd.Series(False, index=df.index)
        trace: list[dict[str, Any]] = []
        for i, item in enumerate(group.conditions):
            child_mask, child_trace = _apply_item(df, item, path=f"{path}.or[{i}]")
            before = int(mask.sum())
            mask = mask | child_mask
            after = int(mask.sum())
            trace.append(
                {
                    "path": f"{path}.or[{i}]",
                    "logic": "or",
                    "added_rows": after - before,
                    "rows_after_union": after,
                    "child": child_trace,
                    "predicate": _describe(item),
                }
            )
        return mask, trace

    # AND — apply high-selectivity (narrowing) predicates first.
    ordered = order_and_children(df, group.conditions)
    mask = pd.Series(True, index=df.index)
    working = df
    trace = []
    for i, item in enumerate(ordered):
        child_mask, child_trace = _apply_item(working, item, path=f"{path}.and[{i}]")
        # child_mask is aligned to `working`; expand to original index via loc.
        aligned = pd.Series(False, index=df.index)
        aligned.loc[working.index] = child_mask.values
        before = int(mask.sum())
        mask = mask & aligned
        after = int(mask.sum())
        working = df.loc[mask]
        pred = _describe(item)
        killed = before > 0 and after == 0
        trace.append(
            {
                "path": f"{path}.and[{i}]",
                "logic": "and",
                "order": i + 1,
                "predicate": pred,
                "estimated_selectivity": estimate_selectivity(df, item),
                "rows_before": before,
                "rows_after": after,
                "dropped": before - after,
                "emptied_result": killed,
                "note": (
                    "This predicate removed every remaining row. "
                    "Combining it with AND hid the rest of the data. "
                    "Consider relaxing it or switching same-column values to OR."
                    if killed
                    else "Applied after more selective predicates so remaining rows stay visible."
                ),
                "child": child_trace,
            }
        )
    return mask, trace


def _apply_item(
    df: pd.DataFrame, item: FilterCondition | FilterGroup, *, path: str
) -> tuple[pd.Series, list[dict[str, Any]]]:
    if isinstance(item, FilterGroup):
        return _apply_group(df, item, path=path)
    return _apply_condition(df, item), []


def _apply_condition(df: pd.DataFrame, cond: FilterCondition) -> pd.Series:
    if cond.column not in df.columns:
        raise ValueError(f"Column '{cond.column}' is not in the frame.")
    series = df[cond.column]
    op = cond.op
    value = cond.value

    if op == "is_null":
        return series.isna()
    if op == "not_null":
        return series.notna()

    if op in {"year_eq", "year_between", "date_before", "date_after"}:
        dates = _as_datetime(series)
        if op == "year_eq":
            year = int(value if not isinstance(value, list) else value[0])
            return dates.dt.year == year
        if op == "year_between":
            lo, hi = _pair(value)
            years = dates.dt.year
            return (years >= int(lo)) & (years <= int(hi))
        if op == "date_before":
            return dates < pd.to_datetime(value, format="mixed")
        return dates > pd.to_datetime(value, format="mixed")

    coerced = coerce_value(series, value)

    if op == "eq":
        return _eq(series, coerced)
    if op == "ne":
        return ~_eq(series, coerced)
    if op == "gt":
        return _numeric(series) > coerced
    if op == "gte":
        return _numeric(series) >= coerced
    if op == "lt":
        return _numeric(series) < coerced
    if op == "lte":
        return _numeric(series) <= coerced
    if op == "between":
        lo, hi = _pair(coerced if isinstance(coerced, list) else value)
        lo_c, hi_c = coerce_value(series, lo), coerce_value(series, hi)
        num = _numeric(series) if pd.api.types.is_numeric_dtype(series) else _as_datetime(series)
        return (num >= lo_c) & (num <= hi_c)
    if op == "in":
        values = coerced if isinstance(coerced, list) else [coerced]
        return _isin(series, values)
    if op == "not_in":
        values = coerced if isinstance(coerced, list) else [coerced]
        return ~_isin(series, values)
    if op == "contains":
        return _as_str(series).str.contains(str(value), case=False, na=False, regex=False)
    if op == "not_contains":
        return ~_as_str(series).str.contains(str(value), case=False, na=False, regex=False)
    if op == "starts_with":
        return _as_str(series).str.startswith(str(value), na=False)
    if op == "ends_with":
        return _as_str(series).str.endswith(str(value), na=False)
    if op == "regex":
        return _as_str(series).str.contains(str(value), case=False, na=False, regex=True)
    raise ValueError(f"Unsupported operator '{op}'")


def _eq(series: pd.Series, value: Any) -> pd.Series:
    if pd.api.types.is_string_dtype(series) or series.dtype == object:
        return _as_str(series).str.casefold() == str(value).casefold()
    if pd.api.types.is_bool_dtype(series):
        return series == bool(value)
    return series == value


def _isin(series: pd.Series, values: list[Any]) -> pd.Series:
    if pd.api.types.is_string_dtype(series) or series.dtype == object:
        folded = {str(v).casefold() for v in values}
        return _as_str(series).str.casefold().isin(folded)
    return series.isin(values)


def _as_str(series: pd.Series) -> pd.Series:
    return series.astype("string")


def _numeric(series: pd.Series) -> pd.Series:
    if pd.api.types.is_numeric_dtype(series):
        return series
    return pd.to_numeric(
        series.astype(str).str.replace(r"[,$]", "", regex=True),
        errors="coerce",
    )


def _as_datetime(series: pd.Series) -> pd.Series:
    if pd.api.types.is_datetime64_any_dtype(series):
        return series
    return pd.to_datetime(series, errors="coerce", format="mixed")


def _pair(value: Any) -> tuple[Any, Any]:
    if isinstance(value, (list, tuple)) and len(value) >= 2:
        return value[0], value[1]
    raise ValueError("between / year_between require a two-element [min, max] value.")


def _describe(item: FilterCondition | FilterGroup) -> dict[str, Any]:
    if isinstance(item, FilterGroup):
        return item.to_dict()
    return item.to_dict()


def human_summary(result: dict[str, Any]) -> str:
    n_in = result["n_input"]
    n_out = result["n_matched"]
    pct = (100.0 * n_out / n_in) if n_in else 0.0
    bits = [f"Matched {n_out:,} of {n_in:,} rows ({pct:.1f}%)."]
    if result.get("empty"):
        bits.append("The result is empty — see the filter trace for which predicate removed the last rows.")
    # Highlight the most destructive AND step.
    def walk(nodes: list[dict[str, Any]]) -> None:
        for node in nodes:
            if node.get("logic") == "and" and node.get("dropped", 0) > 0:
                pred = node.get("predicate") or {}
                bits.append(
                    f"AND step {node.get('order')}: {pred.get('column')} {pred.get('op')} "
                    f"{pred.get('value')} dropped {node['dropped']:,} rows "
                    f"({node['rows_before']:,} → {node['rows_after']:,})."
                )
            child = node.get("child") or []
            if child:
                walk(child)

    walk(result.get("trace") or [])
    return " ".join(bits)


def preview_dict(result: dict[str, Any], *, head: int = 5) -> dict[str, Any]:
    frame: pd.DataFrame = result["frame"]
    payload = {k: v for k, v in result.items() if k != "frame"}
    payload["summary"] = human_summary(result)
    payload["columns"] = [str(c) for c in frame.columns]
    payload["sample_rows"] = json_safe(frame.head(head).to_dict(orient="records"))
    return json_safe(payload)
